infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/CorfuLogReplicationRuntime.java

```diff
@@ -0,0 +1,295 @@
+ * all states in which the leader node on the active cluster can be.
+ *
+ *
+ *                                                       R-LEADER_LOSS
+ *                                             +-------------------------------+
+ *                              ON_CONNECTION  |                               |    ON_CONNECTION_DOWN
+ *                                    UP       |       ON_CONNECTION_DOWN      |       (NON_LEADER)
+ *                                    +----+   |          (R-LEADER)           |
+ *                                    |    |   |   +-----------------------+   |        +-----+
+ *                                    |    |   |   |                       |   |        |     |
+ * +---------------+  ON_CONNECTION  ++----v---v---v--+                  +-+---+--------+-+   |
+ * |               |       UP        |                |  R-LEADER_FOUND  |                <---+
+ * |    WAITING    +---------------->+    VERIFYING   +------------------>                +---+
+ * |      FOR      |                 |     REMOTE     |                  |   NEGOTIATING  |   |  NEGOTIATION_FAILED
+ * |  CONNECTIONS  +<----------------+     LEADER     |                  |                <---+       (ALARM)
+ * |               |  ON_CONNECTION  |                +<-----------+     |                +----+
+ * +---------------+      DOWN       +-^----+---^----++            |     +-------+-----^--+    |
+ *                       (ALL)         |    |   |    |             |             |     |       |
+ *                                     |    |   |    |        R-LEADER_LOSS      |     +-------+
+ *                                     +----+   +----+             |             |  ON_CONNECTION_UP
+ *                              ON_CONNECTION     R-LEADER_NOT     |             |    (NON-LEADER)
+ *                                  DOWN              FOUND        |             |
+ *                                (NOT ALL)                        |     NEGOTIATE_COMPLETE
+ *                                                                 |             |
+ *                                                           ON_CONNECTION       |   ON_CONNECTION_UP
+ *                                                               DOWN            |     (NON-LEADER)
+ *                                                             (R-LEADER)        |      +-----+
+ *                                                                 |             |      |     |
+ *                                                                 |     +-------v------+-+   |
+ *            +---------------+      ALL STATES                    +-----+                <---+
+ *            |               |                                          |                |
+ *            |   STOPPED     <---- L-LEADER_LOSS                        |  REPLICATING   |
+ *            |               |                     SITE FLIP <-----     |                |
```

Comment:
It is possible that while doing the negotiation, there is a leadership change at the standby site.

Reply:
of course, and that will cause the event R_LEADER_LOSS